### ISSUE TITLE
base keyword breaks hash based URIs

### ISSUE THREAD

**Issue by NicoRobertIn** [Wimmics/corese#179]
**Issue Description:**

The parser ruins the URI part passed to the `@base` keyword if this URI part is hash based

**Bug Details:**

When a URI is passed to the base keyword in a turtle file, if this URI ends with a #, then a part of this URI is lost during parsing, ruining all the URIs of the graph using this base

**Steps to Reproduce:**

1. Create a turtle file using a base with a hash based URI and add a triple with a URI using this base. For example this one:

```turtle
@base <https://example.org/route/disappeared#> .

<BrokenURI> a owl:Class .
```

2. Query that base with a construct request that will retrieve that URI, for example:

```sparql
construct {?s ?p ?o} where {?s ?p ?o }
```

**Expected Behavior:**

The broken URI should be `<https://example.org/route/disappeared#BrokenURI>`

**Actual Behavior:**

The following turtle is returned

```turtle
@prefix owl: <http://www.w3.org/2002/07/owl#> .
@prefix ns1: <https://example.org/route/> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .

ns1:BrokenURI a owl:Class .

rdf:type a rdf:Property .
```

From this result we can conclude that the BrokenURI is now `<https://example.org/route/BrokenURI>`, which is different from `<https://example.org/route/disappeared#BrokenURI>`

**Note to Developers:**

This behaviour was tested on Corese python, Corese command and Corese GUI on different computers.

The same URI modification can also be seen with a simple `select * where {?s ?p ?o}` request

**Screenshots/Attachments:**

![image](https://github.com/Wimmics/corese/assets/144037080/d3d629e7-4394-43f0-b601-c0d068f73c6c)
![image](https://github.com/Wimmics/corese/assets/144037080/dd5005ad-2438-46e0-b960-c61456008864)
![image](https://github.com/Wimmics/corese/assets/144037080/4c50381a-0ffd-4c21-9bb6-bd1ab4c86365)

**Comment by FabienGandon** [comment]
I think the base URI must be an absolute URI i.e. an URI with no fragment hence no #

**Comment by frmichel** [comment]
Additional: to fix your problem, you should set

```
@base <https://example.org/route/disappeared> .
<#BrokenURI> a owl:Class .
```

**Comment by MaillPierre** [comment]
The turtle syntax says that ["@base" should be followed by an IRIREF](https://www.w3.org/TR/2014/REC-turtle-20140225/#grammar-production-base).
An [IRIREF must correspond to the following form](https://www.w3.org/TR/2014/REC-turtle-20140225/#grammar-production-IRIREF): `'<' ([^#x00-#x20<>"{}|^`\] | UCHAR)* '>'`
Which I translate as the REGEX: `<([^\x00\-\x20<>"{}|^`\\]|\X)*>`
This regex [validates `<https://raw.githubusercontent.com/acimov-tools/model-test/main/.acimov/custom-tests/shapes.ttl#>`](https://regex101.com/r/LagHHV/1).

Unless my regex is wrong, the Turtle recommendation says that the base URL used in OP's file is correct.

**Comment by MaillPierre** [comment]
@frmichel good catch, I updated the regex by decomposing the UCHAR regex: [https://regex101.com/r/05Bh3v/3](https://regex101.com/r/05Bh3v/3)
It stills validate `<https://example.org/route/disappeared#BrokenURI>`
